gcc/rust/resolve/rust-name-resolution-context.cc

```diff
@@ -82,8 +82,69 @@ BindingLayer::get_source () const
   return source;
 }
 
+Resolver::CanonicalPath
+CanonicalPathRecordRoot::as_path (const NameResolutionContext &)
+{
+  return Resolver::CanonicalPath::create_empty ();
+}
+
+Resolver::CanonicalPath
+CanonicalPathRecordNormal::as_path (const NameResolutionContext &ctx)
+{
+  auto parent_path = get_parent ().as_path (ctx);
+  bool is_crate = parent_path.is_empty ();
+  auto ret = parent_path.append (Resolver::CanonicalPath::new_seg (id, seg));
+  if (is_crate)
+    {
+      auto crate_num = Analysis::Mappings::get ().lookup_crate_num (id);
+      rust_assert (crate_num.has_value ());
+      ret.set_crate_num (*crate_num);
+    }
+  return ret;
+}
+
+Resolver::CanonicalPath
+CanonicalPathRecordLookup::as_path (const NameResolutionContext &ctx)
+{
+  if (!cache)
+    {
+      auto res = ctx.lookup (lookup_id).and_then (
+	[&ctx] (NodeId id) { return ctx.canonical_ctx.get_record_opt (id); });
+
+      if (!res)
+	{
+	  // HACK: use a dummy value
+	  // this should bring us roughly to parity with nr1.0
+	  // since nr1.0 doesn't seem to handle canonical paths for generics
+	  //   quite right anyways
+	  return Resolver::CanonicalPath::new_seg (UNKNOWN_NODEID, "XXX");
```

Comment:
yeah generics are a right pain to handle sometims the paths always default to their non monomorphized version sometimes not but yeah. In order to do it all "correctly" there is going to need to be some mappings between the segment and the conical path segments.. We get away with it for now because it doesnt matter, plus rust will never get a stable ABI and the hash at the end of the managled symbol.

My current feel is that if we can do what we can with canonical paths for now purely in name resolution we get like 80% the way there and its ok.